libyul/optimiser/MemoryStoreRemover.cpp

```diff
@@ -0,0 +1,323 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/optimiser/MemoryStoreRemover.h>
+#include <libyul/optimiser/Semantics.h>
+#include <libyul/optimiser/SSAValueTracker.h>
+
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/backends/wasm/WasmDialect.h>
+
+#include <libyul/Exceptions.h>
+
+#include <libevmasm/Instruction.h>
+#include <libevmasm/SemanticInformation.h>
+
+#include <libsmtutil/SolverInterface.h>
+
+#include <libsolutil/cxx20.h>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+using namespace solidity::util;
+using namespace solidity::evmasm;
+using namespace solidity::smtutil;
+
+namespace
+{
+
+/// Helper function that returns the instruction, if the function
+optional<Instruction> toInstruction(
```

Comment:
Oh wow, this can be useful in a lot of places :)

Reply:
Checked and implemented this refactoring, but only three other places use it.

Reply:
Did this in #11306.